src/micro.c

```diff
@@ -53,10 +55,18 @@ micro_AddService(microService **new_m, natsConnection *nc, microServiceConfig *c
 
     MICRO_CALL(err, _clone_service_config(&m->cfg, cfg));
 
+    // Add the service to the connection.
+    MICRO_CALL(err, _attach_service_to_connection(m->nc, m));
+    MICRO_CALL(err, (m->refs++, NULL));
+
     // Wrap the connection callbacks before we subscribe to anything.
-    MICRO_CALL(err, _wrap_connection_event_callbacks(m));
+    MICRO_CALL(err, micro_ErrorFromStatus(
+                        natsOptions_setMicroCallbacks(m->nc->opts, _on_connection_closed, _on_error)));
```

Comment:
So for instance even if the `nc` connection is valid at the time we enter `micro_AddService`, but happen to be closed during execution of this function, it will leave things in a bad state. If service is part of a connection, I think it should be handled by the connection, even if `micro_X` calls are just a proxy to internal `natsConn_X` calls. Have you thought that way?

Reply:
I did, in fact I was VERY inclined to make them "1st class citizens" in `natsConnection`, but not for this PR...  This all really came out from trying to fix a single race condition on that global services table I had before, wrong locking order in 1 place.

OT: In light of the `orbit` movement I've been contemplating adding simple "internal use" extensibility to both Connection and Subscription; for data (like a hashmap) and callbacks (say, with a `bool lock` parameter). My initial microservices implementation was naive, it is still marked experimental, and I would really love to externalize and re-factor it to orbit before it is final.

Reply:
I moved all "attach to connection" functionality into the func, and added a `natsConn_Lock` around + connection validation. Thanks for pointing it out. After the service is added to connection, it should unwind "normally" if it's terminated.